torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -3382,16 +3389,75 @@ at::Tensor interpolate(
     }
   }
 
+  float scale_factors_1 = -1.0f;
+  float scale_factors_2 = -1.0f;
+  float scale_factors_3 = -1.0f;
+
+  if(!scale_factors.isNone() && use_scale_factor == c10::nullopt) {
+    use_scale_factor = false;
+    bool warn_use_scale_factor = false;
+
+    if (scale_factors.isDouble()) {
+      if(_is_float_value(static_cast<float>(scale_factors.toDouble()))){
+        warn_use_scale_factor = true;
+      }
+    } else if (scale_factors.isDoubleList()) {
+      auto scale_factors_list_ref = scale_factors.toDoubleListRef();
+      std::vector<double> scale_factors_vec(scale_factors_list_ref.begin(), scale_factors_list_ref.end());
+      auto scale_factors_list = c10::impl::toList(scale_factors_vec);
+
+      if(_is_float_value(static_cast<float>(scale_factors_list[0]))){
+        warn_use_scale_factor = true;
+      }
+      else if (scale_factors_list.size() >= 2 &&
+               _is_float_value(static_cast<float>(scale_factors_list[1]))) {
+                 warn_use_scale_factor = true;
+      }
+      else if (scale_factors_list.size() >= 3 &&
+          _is_float_value(static_cast<float>(scale_factors_list[2]))) {
+            warn_use_scale_factor = true;
+      }
+    }
+
+    if(warn_use_scale_factor) {
+      AT_WARN(
+        "The default behavior for interpolate/upsample with float scale_factor will change "
+        "in 1.5.0 to align with other frameworks/libraries, and use scale_factor directly, "
+        "instead of relying on the computed output size. "
+        "If you wish to keep the old behavior, please set use_scale_factor=False. "
+        "See the documentation of nn.Upsample for details.");
+    }
+  }
+
+  if(use_scale_factor) {
+    if (scale_factors.isDouble()) {
+      scale_factors_1 = static_cast<float>(scale_factors.toDouble());
+      scale_factors_2 = static_cast<float>(scale_factors.toDouble());
+      scale_factors_3 = static_cast<float>(scale_factors.toDouble());
+    } else if (scale_factors.isDoubleList()) {
+      auto scale_factors_list_ref = scale_factors.toDoubleListRef();
```

Comment:
All of the separate handling of the 1, 2, 3 cases is really showing that we should use a `float[]`, the logic would be much simpler.